plugins/query-insights/src/main/java/org/opensearch/plugin/insights/core/listener/QueryInsightsListener.java

```diff
@@ -0,0 +1,141 @@
+            );
+        this.setEnabled(MetricType.LATENCY, clusterService.getClusterSettings().get(TOP_N_LATENCY_QUERIES_ENABLED));
+        this.queryInsightsService.setTopNSize(clusterService.getClusterSettings().get(TOP_N_LATENCY_QUERIES_SIZE));
+        this.queryInsightsService.setWindowSize(clusterService.getClusterSettings().get(TOP_N_LATENCY_QUERIES_WINDOW_SIZE));
+    }
+
+    /**
+     * Enable or disable metric collection for {@link MetricType}
+     *
+     * @param metricType {@link MetricType}
+     * @param enabled boolean
+     */
+    public void setEnabled(MetricType metricType, boolean enabled) {
+        this.queryInsightsService.enableCollection(metricType, enabled);
+
+        // disable QueryInsightsListener only if collection for all metrics are disabled.
+        if (!enabled) {
+            for (MetricType t : MetricType.allMetricTypes()) {
+                if (this.queryInsightsService.isCollectionEnabled(t)) {
+                    return;
+                }
+            }
+            super.setEnabled(false);
+        } else {
+            super.setEnabled(true);
+        }
+    }
+
+    @Override
+    public boolean isEnabled() {
+        return super.isEnabled();
+    }
+
+    @Override
+    public void onPhaseStart(SearchPhaseContext context) {}
+
+    @Override
+    public void onPhaseEnd(SearchPhaseContext context, SearchRequestContext searchRequestContext) {}
+
+    @Override
+    public void onPhaseFailure(SearchPhaseContext context) {}
+
+    @Override
+    public void onRequestStart(SearchRequestContext searchRequestContext) {}
+
+    @Override
+    public void onRequestEnd(SearchPhaseContext context, SearchRequestContext searchRequestContext) {
+        SearchRequest request = context.getRequest();
+        try {
+            Map<MetricType, Measurement<? extends Number>> measurements = new HashMap<>();
+            if (queryInsightsService.isCollectionEnabled(MetricType.LATENCY)) {
+                measurements.put(
+                    MetricType.LATENCY,
+                    new Measurement<>(
+                        MetricType.LATENCY.name(),
+                        TimeUnit.NANOSECONDS.toMillis(System.nanoTime() - searchRequestContext.getAbsoluteStartNanos())
+                    )
+                );
+            }
+            Map<Attribute, Object> attributes = new HashMap<>();
+            attributes.put(Attribute.SEARCH_TYPE, request.searchType().toString().toLowerCase(Locale.ROOT));
+            attributes.put(Attribute.SOURCE, request.source().toString(FORMAT_PARAMS));
+            attributes.put(Attribute.TOTAL_SHARDS, context.getNumShards());
+            attributes.put(Attribute.INDICES, request.indices());
+            attributes.put(Attribute.PHASE_LATENCY_MAP, searchRequestContext.phaseTookMap());
+            SearchQueryRecord record = new SearchQueryRecord(request.getOrCreateAbsoluteStartMillis(), measurements, attributes);
+            queryInsightsService.addRecord(record);
```

Comment:
is it a blocking call?

Reply:
This is the `addRecord ` function we discussed about, Internally it will submit the job to the threadpool to handle the data asynchronously with the query insights executor. https://github.com/opensearch-project/OpenSearch/pull/11903/files#diff-529ce87837ab403eb338746abbe3b7fc055c823e2b2fd039e56d193d2f01303bR112